File: go/observer/observer_test.go
package observer

import (
	"github.com/stretchr/testify/assert"
	"sync"
	"testing"
)

var got = make([][]byte, 0) // For storing received messages to assert later
var wg sync.WaitGroup       // For waiting goroutines to finish in test

// SubscriberImpl implements Subscriber interface
type SubscriberImpl struct {
}

// Update is called when a message is received
func (s SubscriberImpl) Update(m Message) {
	got = append(got, m.Data)
	defer wg.Done()
}

func TestObserver(t *testing.T) {
	sub1 := SubscriberImpl{}           // Create a new subscriber
	sub2 := SubscriberImpl{}           // Create another
	broker := NewPublisher(sub1, sub2) // Create a broker with 2 subscribers

	wg.Add(6) // We will send 3 messages to 2 subscribers so 6
	for i := 0; i < 3; i++ {
		go broker.Pub(Message{Data: []byte("Hello")}) // Publish messages
	}
	wg.Wait() // Wait for goroutines to finish

	assert.Len(t, got, 6)
	for _, v := range got {
		assert.Equal(t, []byte("Hello"), v)
	}
}

File: typescript/src/observer/observer.test.ts
import { Publisher, Subscriber } from "./observer"

const sub1: Subscriber = {
  update: jest.fn(),
}

const sub2: Subscriber = {
  update: jest.fn(),
}

const sub3: Subscriber = {
  update: jest.fn(),
}

test("Observer", () => {
  const broker = new Publisher(sub1, sub2)
  broker.sub(sub3)
  broker.pub({ data: "Hello" })

  expect(sub1.update).toHaveBeenCalledWith({ data: "Hello" })
  expect(sub2.update).toHaveBeenCalledWith({ data: "Hello" })
  expect(sub3.update).toHaveBeenCalledWith({ data: "Hello" })
})
